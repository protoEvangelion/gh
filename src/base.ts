/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as configs from './configs'
import { safeReaddir } from './fp'
import * as Future from 'fluture'

// -- Config -------------------------------------------------------------------

export function clone(o) {
    return JSON.parse(JSON.stringify(o))
}

// -- Utils --------------------------------------------------------------------

/**
 * Returns files in a folder path that match a given patter
 */
export function find(
    dirPath: string,
    optPattern = /.*/
): Future.FutureInstance<NodeJS.ErrnoException, string[]> {
    return safeReaddir(dirPath).map(dirs => {
        return dirs.filter(file => {
            return optPattern.test(file)
        })
    })
}

export function getUser() {
    return configs.getConfig().github_user || process.env.GH_USER
}

// Export some config methods to allow plugins to access them
export const getConfig = configs.getConfig
export const writeGlobalConfig = configs.writeGlobalConfig
