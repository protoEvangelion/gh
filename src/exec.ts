/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as childProcess from 'child_process'
import * as logger from './logger'

export function spawnSync(cmd, args, options?: object) {
    let exec

    logger.debug(`spawnSync: ${cmd} ${args.join(' ')}`)

    exec = childProcess.spawnSync(cmd, args, options)

    if (exec.error && exec.error.code === 'ENOENT' && process.platform === 'win32') {
        logger.debug("Invoking patched sapwnSync due to Windows' libuv bug")
        exec = childProcess.spawnSync(`${cmd}.cmd`, args, options)
    }

    return {
        stdout: exec.stdout.toString().trim(),
        stderr: exec.stderr.toString().trim(),
        status: exec.status,
    }
}

export function spawnSyncStream(cmd, args, options?: any) {
    let proc
    let err

    if (!options) {
        options = {}
    }

    options.stdio = ['pipe', process.stdout, process.stderr]

    logger.debug(`spawnSyncStream: ${cmd} ${args.join(' ')}`)

    proc = childProcess.spawnSync(cmd, args, options)

    if (proc.status !== 0) {
        err = new Error()
        err.code = proc.status
        err.message = `Child process terminated with error code ${err.code}`

        throw err
    }

    return proc
}

export function execSync(cmd, options?) {
    if (!options) {
        options = {}
    }

    logger.debug(`execSync: ${cmd}`)

    options.stdio = ['pipe', process.stdout, process.stderr]

    return childProcess.execSync(cmd, options)
}

export function execSyncInteractiveStream(cmd, options?) {
    if (!options) {
        options = {}
    }

    logger.debug(`execSyncInteractiveStream: ${cmd}`)

    options.stdio = 'inherit'

    return childProcess.execSync(cmd, options)
}
