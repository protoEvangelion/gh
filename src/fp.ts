/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as S from 'sanctuary'
import * as Future from 'fluture'
import * as fs from 'fs'
import * as which from 'which'
import * as $ from 'sanctuary-def'

/* SAFE UTILS */
export const safeWhich = Future.encase(which.sync)
export const safeRealpath = Future.encase(fs.realpathSync)

export function safeImport(
    fileName: string
): Future.FutureInstance<NodeJS.ErrnoException, object | string> {
    let moduleObj = null

    try {
        moduleObj = require(fileName)
    } catch (e) {
        return e.code === 'MODULE_NOT_FOUND' ? Future.of(fileName) : Future.reject(e)
    }

    return Future.of(moduleObj)
}

export const safeReaddir = Future.encaseN<any, string[], string>(fs.readdir)

/* TRANSFORMATIONS */
// Allows you to take a Maybe returning function and make it an Either returning function
export function maybeFnToEither(monadReturningFunction) {
    return function convertMaybe(val) {
        const result = monadReturningFunction(val)

        return S.isJust(result) ? S.Right(result.value) : S.Left(null)
    }
}

/**
 * Concats two strings
 * @return {Future}
 */
export const prepend = (a: string) => (b: string) => {
    const argsAreStrings = typeof a === 'string' && typeof b === 'string'

    return argsAreStrings ? Future.of(a + b) : Future.reject('Both args should be strings')
}

/* TYPE CHECKING UTILS */
export const isObject = S.is($.Object)
export const isString = S.is($.String)

/* LOGGING UTILS */
export const l = x => {
    console.log('!!!!!!!', x)
    return x
}

// returns a future
export const lf = x => {
    console.log('-------', x)
    return Future.of(x)
}
