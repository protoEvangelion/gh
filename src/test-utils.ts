/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import { execSync } from 'child_process'
import { isArray, isObject, isPlainObject, map, mapValues, upperFirst } from 'lodash'
import * as nock from 'nock'
import * as zlib from 'zlib'

export function runCmd(cmd: string, env?: boolean) {
    try {
        const customEnv = env ? { env: { ...process.env } } : {}

        var result = execSync(cmd, { cwd: process.cwd(), ...customEnv })
    } catch (error) {
        throw new Error(error.output.toString())
    }

    return result.toString()
}

const nockBack = nock.back

export function prepareTestFixtures(cmdName, argv) {
    let id = 0

    // These should only include the flags that you need for e2e tests
    const cmds = [
        {
            name: 'Help',
            flags: [],
        },
        {
            name: 'Issue',
            flags: ['--comment', '--new', '--open', '--close', '--search', '--assign'],
        },
        {
            name: 'PullRequest',
            flags: [
                '--detailed',
                '--info',
                '--fetch',
                '--fwd',
                '--comment',
                '--open',
                '--close',
                '--draft',
                '--submit',
            ],
        },
        {
            name: 'Gists',
            flags: ['--new', '--fork', '--delete'],
        },
        {
            name: 'Milestone',
            flags: ['--list'],
        },
        {
            name: 'Notifications',
        },
        {
            name: 'Repo',
            flags: ['--label', '--list', '--new', '--fork', '--delete', '--search'],
        },
        {
            name: 'User',
            flags: ['--login', '--logout', '--whoami'],
        },
        {
            name: 'Version',
            flags: ['--version'],
        },
    ].filter(cmd => cmd.name === cmdName)

    const newCmdName = formatCmdName(cmds[0], argv)

    if (!newCmdName) {
        return () => {}
    }

    nock.disableNetConnect()
    nockBack.fixtures = `${process.cwd()}/__tests__/nockFixtures`
    nockBack.setMode('record')

    const nockPromise = nockBack(`${newCmdName}.json`, {
        before,
        afterRecord,
    })

    return () =>
        nockPromise
            .then(({ nockDone }) => nockDone())
            .catch(err => {
                throw new Error(`Nock ==> ${err}`)
            })

    /* --- Normalization Functions --- */

    function normalize(value, key) {
        if (!value) return value

        if (isPlainObject(value)) {
            return mapValues(value, normalize)
        }

        if (isArray(value) && isPlainObject(value[0])) {
            return map(value, normalize)
        }

        if (key.includes('token')) {
            return '234lkj23l4kj234lkj234lkj234lkj23l4kj234l'
        }

        if (key.includes('_at')) {
            return '2017-10-10T16:00:00Z'
        }

        if (key.includes('_count')) {
            return 42
        }

        if (key.includes('id')) {
            return 1000 + id++
        }

        if (key.includes('node_id')) {
            return 'MDA6RW50aXR5MQ=='
        }

        if (key.includes('url')) {
            return value.replace(/[1-9][0-9]{2,10}/, '000000001')
        }

        return value
    }

    function decodeBuffer(fixture) {
        const response = isArray(fixture.response) ? fixture.response.join('') : fixture.response

        if (!isObject(response)) {
            try {
                // Decode the hex buffer that nock made
                const decoded = Buffer.from(response, 'hex')
                var unzipped = zlib.gunzipSync(decoded).toString('utf-8')
            } catch (err) {
                throw new Error(`Error decoding nock hex:\n${err}`)
            }
        }

        return JSON.parse(unzipped)
    }

    // This only executes when first recording the request, but not on subsequent requests
    function afterRecord(fixtures) {
        const normalizedFixtures = fixtures.map(fixture => {
            const isGzipped = fixture.rawHeaders.includes('gzip')
            let res = fixture.response

            if (fixture.body.note) {
                fixture.body.note = 'Hello from the inside!'
            }

            fixture.path = stripAccessToken(fixture.path)
            fixture.rawHeaders = fixture.rawHeaders.map(header => stripAccessToken(header))

            if (isGzipped) {
                res = decodeBuffer(fixture)
            }

            if (isArray(res)) {
                res = res.slice(0, 3).map(res => {
                    return mapValues(res, normalize)
                })
            } else {
                res = mapValues(res, normalize)
            }

            if (isGzipped) {
                try {
                    // Re-gzip to keep the octokittens happy
                    const stringified = JSON.stringify(res)
                    var zipped = zlib.gzipSync(stringified)
                } catch (err) {
                    throw new Error(`Error re-gzipping nock ==> ${err}`)
                }
            }

            fixture.response = zipped || res

            return fixture
        })

        return normalizedFixtures
    }

    function stripAccessToken(header) {
        return header.includes('access_token')
            ? header.replace(/access_token(.*?)(&|$)/gi, '')
            : header
    }

    function before(scope) {
        scope.filteringPath = () => stripAccessToken(scope.path)
        scope.filteringRequestBody = (body, aRecordedBody) => {
            if (body.includes('note')) {
                body = JSON.parse(body)
                body.note = aRecordedBody.note

                return JSON.stringify(body)
            }

            return body
        }
    }
}

function formatCmdName(cmd, argv) {
    if (argv.length === 1) {
        return cmd.name
    }

    return cmd.flags.reduce((flagName, current) => {
        if (flagName) {
            return flagName
        }

        if (argv.includes(current)) {
            return concatUpper(cmd.name, current.slice(2))
        }
    }, null)
}

function concatUpper(one, two) {
    return `${one}${upperFirst(two)}`
}
