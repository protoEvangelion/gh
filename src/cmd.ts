/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

// -- Requires -------------------------------------------------------------------------------------

import * as Future from 'fluture'
import { env as flutureEnv } from 'fluture-sanctuary-types'
import * as fs from 'fs'
import produce, { setAutoFreeze } from 'immer'
import * as nopt from 'nopt'
import * as path from 'path'
import * as R from 'ramda'
import { create, env } from 'sanctuary'
import * as updateNotifier from 'update-notifier'
import { getConfig, getUser } from './base'
import * as logger from './logger'
import {
    createGlobalConfig,
    getGlobalPackageJson,
    getUserHomePath,
    addPluginConfig,
} from './configs'
import { prepend, safeReaddir, safeImport, safeRealpath, safeWhich } from './fp'
import * as git from './git'
import { getGitHubInstance } from './github'

const testing = process.env.NODE_ENV === 'testing'

// Make Fluture Play nicely with Sanctuary
const S = create({ checkTypes: true, env: env.concat(flutureEnv) })

// Allow mutation of options when not testing
// https://immerjs.github.io/immer/docs/freezing
setAutoFreeze(testing)

Future.debugMode(testing)

interface CommandInterface {
    name: string
    isPlugin?: boolean
    DETAILS: {
        alias: string
        description: string
        commands: string
        options: object
        shorthands: object
    }
    run: (options?: any, done?: any) => {}
}

interface PluginInterface extends CommandInterface {
    new (any): CommandInterface
}

// -- Utils ----------------------------------------------------------------------------------------

interface Args {
    cooked?: string[]
    remain?: string[]
}

/**
 * Figure out if cmd is either the Version of Help cmd
 */
export function tryResolvingByHelpOrVersion({ cooked, remain }: Args = {}): Future.FutureInstance<
    string,
    string
> {
    let cmdName = null

    const isVersionCmd = cooked[0] === '--version' || cooked[0] === '-v'
    const isHelpCmd = !remain.length || cooked.includes('-h') || cooked.includes('--help')

    if (isVersionCmd) {
        cmdName = 'version'
    } else if (isHelpCmd) {
        cmdName = 'help'
    }

    return cmdName ? Future.of(cmdName) : Future.reject(remain[0])
}

/**
 * Builds out the absolute path of the non plugin cmd
 */
function buildFilePath(filename: string): string {
    const commandDir = path.join(__dirname, 'cmds')

    const fullFileName = filename.includes('.') ? filename : `${filename}`
    const absolutePath = path.join(commandDir, fullFileName)

    return absolutePath
}

type TryResolvingByPlugin = (a: string) => Future.FutureInstance<NodeJS.ErrnoException, string>
/**
 * Try to determine if cmd passed in is a plugin
 */
export const tryResolvingByPlugin: TryResolvingByPlugin = R.pipeK(
    prepend('gh-'),
    safeWhich,
    safeRealpath
)

/**
 * Checks if cmd is a valid alias
 */
export function tryResolvingByAlias(name: string): Future.FutureInstance<string, string> {
    const cmdDir = path.join(__dirname, 'cmds')

    return safeReaddir(cmdDir)
        .chain(filterFiles)
        .chainRej(() => Future.reject(name))

    function filterFiles(files: string[]): Future.FutureInstance<string, string> {
        const cmdFileName = files.filter((file: string) => {
            return file.startsWith(name[0]) && file.includes(name[1])
        })[0]

        return cmdFileName ? Future.of(cmdFileName) : Future.reject(name)
    }
}

// Some plugins have the Impl prop housing the main class
// For backwards compat, we will flatten it if it exists
function flattenIfImpl(obj) {
    return obj.Impl ? obj.Impl : obj
}

export function loadCommand(
    args: Args
): Future.FutureInstance<NodeJS.ErrnoException, CommandInterface> {
    return tryResolvingByHelpOrVersion(args)
        .chainRej(tryResolvingByAlias)
        .map(buildFilePath)
        .chainRej(tryResolvingByPlugin)
        .chain(safeImport)
        .map(flattenIfImpl)
}

function getCommand(
    args: string[]
): Future.FutureInstance<{ value: string }, { value: CommandInterface }> {
    /**
     * nopt function returns:
     *
     * remain: The remaining args after all the parsing has occurred.
     * original: The args as they originally appeared.
     * cooked: The args after flags and shorthands are expanded.
     */
    const parsed = nopt(args)
    const remain = parsed.argv.remain
    const module = remain[0]

    const Command = loadCommand(parsed.argv)

    return Command.fold(() => S.Left(`Cannot find module ${module}`), S.Right)
}

function notifyVersion(): void {
    const notifier = updateNotifier({ pkg: getGlobalPackageJson() })

    if (notifier.update) {
        notifier.notify()
    }
}

export async function buildOptions(args, cmdName) {
    const options = produce(args, async draft => {
        const config = getConfig()

        // Gets 2nd positional arg (`gh pr 1` will return 1)
        const secondArg = [draft.argv.remain[1]]
        const remote = draft.remote || config.default_remote
        const remoteUrl = git.getRemoteUrl(remote)

        if (cmdName !== 'Help' && cmdName !== 'Version') {
            // We don't want to boot up Ocktokit if user just wants help or version
            draft.GitHub = await getGitHubInstance()
        }

        // default the page size to 30
        draft.allPages = config.page_size === ''
        draft.pageSize = config.page_size || 30
        draft.config = config
        draft.remote = remote
        draft.number = draft.number || secondArg
        draft.loggedUser = getUser()
        draft.remoteUser = git.getUserFromRemoteUrl(remoteUrl)
        draft.repo = draft.repo || git.getRepoFromRemoteURL(remoteUrl)
        draft.currentBranch = git.getCurrentBranch()
        draft.github_host = config.github_host
        draft.github_gist_host = config.github_gist_host

        if (!draft.user) {
            if (args.repo || args.all) {
                draft.user = draft.loggedUser
            } else {
                draft.user = process.env.GH_USER || draft.remoteUser || draft.loggedUser
            }
        }

        /**
         * Checks if there are aliases in your .gh.json file.
         * If there are aliases in your .gh.json file, we will attempt to resolve the user, PR forwarder or PR submitter to your alias.
         */
        if (config.alias) {
            draft.fwd = config.alias[draft.fwd] || draft.fwd
            draft.submit = config.alias[draft.submit] || draft.submit
            draft.user = config.alias[draft.user] || draft.user
        }

        draft.userRepo = logger.colors.green(`${draft.user}/${draft.repo}`)
    })

    return options
}

/* !! IMPURE CALLING CODE !! */
export async function run() {
    process.env.GH_PATH = path.join(__dirname, '../')

    if (!fs.existsSync(getUserHomePath())) {
        createGlobalConfig()
    }

    notifyVersion()

    getCommand(process.argv).fork(
        errMsg => console.log(errMsg),
        async ({ value: Command }) => {
            const args = getAvailableArgsOnCmd(Command)
            let cmdDoneRunning = null

            if (testing) {
                const { prepareTestFixtures } = await import('./test-utils')

                // Enable mock apis for e2e's
                cmdDoneRunning = prepareTestFixtures(Command.name, args.argv.cooked)
            }

            const options = await buildOptions(args, Command.name)

            // Maintain backwards compat with plugins implemented as classes
            if (typeof Command === 'function') {
                const Plugin: PluginInterface = Command

                addPluginConfig(Plugin.name)

                await new Plugin(options).run(cmdDoneRunning)
            } else {
                await Command.run(options, cmdDoneRunning)
            }
        }
    )
}

/**
 * If you run `gh pr 1 -s node-gh --remote=origin --user protoEvangelion`, nopt will return
 *
 *   {
 *     remote: 'origin',
 *     submit: 'node-gh',
 *     user: 'protoEvangelion',
 *     argv: {
 *         original: ['pr', '1', '-s', 'pr', 'node-gh', '--remote', 'origin', '--user', 'protoEvangelion'],
 *         remain: ['pr', '1'],
 *         cooked: ['pr', '1', '--submit', 'node-gh', '--remote', 'origin', '--user', 'protoEvangelion'],
 *     },
 *   }
 *
 * Historically we passed every arg after 2nd arg (gh pr 1 -s user; everything after 'pr')
 * and all parsed options to each cmd's payload function to figure out positional args and allow for neat shortcuts like:
 * gh is 'new issue' 'new issue description'
 */
function getAvailableArgsOnCmd(Command: CommandInterface) {
    return nopt(Command.DETAILS.options, Command.DETAILS.shorthands, process.argv, 2)
}
