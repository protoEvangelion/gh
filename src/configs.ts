/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as fs from 'fs'
import { cloneDeep } from 'lodash'
import * as path from 'path'
import * as R from 'ramda'
import * as S from 'sanctuary'
import * as userhome from 'userhome'
import * as which from 'which'
import * as exec from './exec'
import * as logger from './logger'
import { safeImport } from './fp'

export const PLUGINS_PATH_KEY = 'plugins_path'

const testing = process.env.NODE_ENV === 'testing'

/* Refactored FP Functions */

const safeWhich = S.encase(which.sync)
const safeRealpath = S.encase(fs.realpathSync)

export const getPluginPath = R.pipeK(
    safeWhich,
    safeRealpath
)

// TODO merge with getPlugin fn in cmd.ts
export const getPlugin = R.pipeK(
    S.prepend('gh-'),
    getPluginPath,
    safeImport
)

/* ----------------------- */

// function concatError(leftMonad, errorMessage) {
//     const leftMonadContainsError = leftMonad && leftMonad.isLeft() && leftMonad.value

//     return leftMonadContainsError ? leftMonad.map(mapError) : Left(errorMessage)

//     function mapError(prevErrorMessage) {
//         return `${prevErrorMessage}\n${errorMessage}`
//     }
// }

// -- Config -------------------------------------------------------------------

export function getNodeModulesGlobalPath() {
    try {
        var { stdout } = exec.spawnSync('npm', ['root', '-g'])
    } catch (err) {
        logger.warn(`Can't resolve plugins directory path.\n${err}`)
    }

    return stdout
}

export function getGlobalPackageJson() {
    const configFile = fs.readFileSync(path.join(__dirname, '../package.json'))

    return JSON.parse(configFile.toString())
}

export function getDefaultConfigPath() {
    return path.join(__dirname, '../default.gh.json')
}

export function getProjectConfigPath() {
    return path.join(process.cwd(), '.gh.json')
}

export function getUserHomePath() {
    return userhome('.gh.json')
}

function resolveGHConfigs() {
    const globalConfig = getGlobalConfig()
    let projectConfig
    const result = {}

    try {
        projectConfig = JSON.parse(fs.readFileSync(getProjectConfigPath()).toString())

        Object.keys(globalConfig).forEach(key => {
            result[key] = globalConfig[key]
        })

        Object.keys(projectConfig).forEach(key => {
            result[key] = projectConfig[key]
        })

        return result
    } catch (e) {
        logger.debug(e.message)

        if (e.code !== 'MODULE_NOT_FOUND' && e.code !== 'ENOENT') {
            throw e
        }

        return globalConfig
    }
}

export function getConfig() {
    const config = resolveGHConfigs()

    const protocol = `${config.api.protocol}://`
    const is_enterprise = config.api.host !== 'api.github.com'

    if (config.github_host === undefined) {
        config.github_host = `${protocol}${is_enterprise ? config.api.host : 'github.com'}`
    }
    if (config.github_gist_host === undefined) {
        config.github_gist_host = `${protocol}${
            is_enterprise ? `${config.api.host}/gist` : 'gist.github.com'
        }/`
    }

    return config
}

export function getGlobalConfig() {
    const configPath = getUserHomePath()
    const defaultPath = getDefaultConfigPath()

    if (!fs.existsSync(configPath)) {
        createGlobalConfig()
    }

    return JSON.parse(fs.readFileSync(testing ? defaultPath : configPath).toString())
}

export function removeGlobalConfig(key) {
    var config = getGlobalConfig()

    delete config[key]

    saveJsonConfig(getUserHomePath(), config)
}

export function createGlobalConfig() {
    saveJsonConfig(
        getUserHomePath(),
        JSON.parse(fs.readFileSync(getDefaultConfigPath()).toString())
    )
}

export function writeGlobalConfig(jsonPath, value) {
    const config = getGlobalConfig()
    let i
    let output
    let path
    let pathLen

    path = jsonPath.split('.')
    output = config

    for (i = 0, pathLen = path.length; i < pathLen; i++) {
        output[path[i]] = config[path[i]] || (i + 1 === pathLen ? value : {})
        output = output[path[i]]
    }

    saveJsonConfig(getUserHomePath(), config)
}

export function saveJsonConfig(path, object) {
    fs.writeFileSync(path, JSON.stringify(object, null, 4))
}

export function writeGlobalConfigCredentials(user, token, path): void {
    const configPath = path || getUserHomePath()

    let config
    if (fs.existsSync(configPath)) {
        config = JSON.parse(fs.readFileSync(configPath).toString())
    } else {
        config = JSON.parse(fs.readFileSync(getDefaultConfigPath()).toString())
    }

    logger.log(`Writing GH config data: ${configPath}`)

    try {
        config.github_user = user
        config.github_token = token

        saveJsonConfig(configPath, config)
    } catch (err) {
        throw new Error(`Error writing credentials to global config\n${err}`)
    }

    logger.log('Authentication succeed. Token written to global config.')
}

// -- Plugins ------------------------------------------------------------------

export function addPluginConfig(plugin) {
    try {
        const pluginConfig = require(path.join(
            getNodeModulesGlobalPath(),
            `gh-${plugin}`,
            'gh-plugin.json'
        ))

        const config = getGlobalConfig()
        const configHooks = cloneDeep(config.hooks)
        const pluginHooks = cloneDeep(pluginConfig.hooks)

        if (config.plugins[plugin] && !config.plugins[plugin]['hooks_installed']) {
            Object.keys(pluginHooks).forEach(cmd => {
                Object.keys(pluginHooks[cmd]).forEach(hook => {
                    configHooks[cmd][hook].before = [
                        ...configHooks[cmd][hook].before,
                        ...pluginHooks[cmd][hook].before,
                    ]

                    configHooks[cmd][hook].after = [
                        ...configHooks[cmd][hook].after,
                        ...pluginHooks[cmd][hook].after,
                    ]
                })
            })

            if (!testing) {
                logger.log(
                    logger.colors.yellow(
                        `Copying over ${plugin} plugin hooks to your .gh.json hooks.`
                    )
                )

                try {
                    config.hooks = configHooks
                    config.plugins[plugin]['hooks_installed'] = true

                    saveJsonConfig(getUserHomePath(), config)
                } catch (err) {
                    logger.error(`Error writing ${plugin} hooks to .gh.json config.\n${err}`)
                }

                logger.log(logger.colors.green('Copy successful.\n'))
            }
        }
    } catch (e) {
        if (e.code !== 'MODULE_NOT_FOUND') {
            throw e
        }
    }
}

export function getPlugins() {
    const pluginsPath = getNodeModulesGlobalPath()

    if (pluginsPath === '') {
        return []
    }

    try {
        var plugins = fs.readdirSync(pluginsPath).filter(plugin => {
            return plugin.substring(0, 3) === 'gh-'
        })
    } catch (err) {
        logger.warn(`Can't read plugins directory.\n${err}`)
    }

    return plugins
}

export function pluginHasConfig(pluginName) {
    return Boolean(getConfig().plugins[pluginName])
}
