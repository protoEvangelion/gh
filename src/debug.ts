#!/usr/bin/env node

/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import { run } from './cmd'

const verbose = process.argv.indexOf('--verbose') !== -1
const insane = process.argv.indexOf('--insane') !== -1

process.on('unhandledRejection', r => console.log(r))

if (verbose || insane) {
    process.env.GH_VERBOSE = 'true'
}

if (insane) {
    process.env.GH_VERBOSE_INSANE = 'true'
}

run()
