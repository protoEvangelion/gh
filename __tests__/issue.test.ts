/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import { runCmd } from '../src/test-utils'

describe('E2E: Issues Module Test', () => {
    it('List Issues `gh is`', done => {
        expect(runCmd('gh is')).toMatchSnapshot()
        done()
    })

    it(`Assign Issues \`gh is --assign -A ${process.env.GH_USER} --number 1\``, done => {
        expect(runCmd(`gh is --assign -A ${process.env.GH_USER} --number 1`)).toMatchSnapshot()
        done()
    })

    it('Open issue in browser `gh is 55 --browser`', done => {
        expect(runCmd('gh is 55 --browser')).toMatchSnapshot()
        done()
    })

    it('Create a new issue `gh is -N -t "Node GH rocks!" -m "msg" -L bug,question,test`', done => {
        expect(
            runCmd(`gh is -N -t "Node GH rocks!" -m "msg" -L bug,question,test`)
        ).toMatchSnapshot()
        done()
    })

    it('Comment on Issues `gh is 1 -c "comment"`', done => {
        expect(runCmd(`gh is 1 -c "comment"`)).toMatchSnapshot()
        done()
    })

    it('Close Issue `gh is --close --number 7 --number 10`', done => {
        expect(runCmd(`gh is --close --number 7 --number 10`)).toMatchSnapshot()
        done()
    })

    it('Open Issue `gh is --open --number 7 --number 10`', done => {
        expect(runCmd(`gh is --open --number 7 --number 10`)).toMatchSnapshot()
        done()
    })

    it('Search Issues `gh is -s hi`', done => {
        expect(runCmd(`gh is -s 'hi'`)).toMatchSnapshot()
        done()
    })
})
