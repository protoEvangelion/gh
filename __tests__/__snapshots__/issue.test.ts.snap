// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E: Issues Module Test Assign Issues \`gh is --assign -A protoEvangelion --number 1\` 1`] = `
"Assigning issue #1 on protoEvangelion/gh to protoEvangelion
https://github.com/protoEvangelion/gh/issues/1
"
`;

exports[`E2E: Issues Module Test Close Issue \`gh is --close --number 7 --number 10\` 1`] = `
"Closing issue 7 on protoEvangelion/gh
https://github.com/protoEvangelion/gh/issues/7
Closing issue 10 on protoEvangelion/gh
https://github.com/protoEvangelion/gh/issues/10
"
`;

exports[`E2E: Issues Module Test Comment on Issues \`gh is 1 -c "comment"\` 1`] = `
"Adding comment on issue #1 on protoEvangelion/gh
https://github.com/protoEvangelion/gh/issues/1#issuecomment-000000001
"
`;

exports[`E2E: Issues Module Test Create a new issue \`gh is -N -t "Node GH rocks!" -m "msg" -L bug,question,test\` 1`] = `
"Creating a new issue on protoEvangelion/gh
https://github.com/protoEvangelion/gh/issues/000000001
[hook] gh is --browser  --user protoEvangelion --repo gh --number 132
"
`;

exports[`E2E: Issues Module Test List Issues \`gh is\` 1`] = `
"Listing open issues on protoEvangelion/gh
#99 notification @protoEvangelion (in 4 years)

#98 pr title2 @protoEvangelion (in 4 years)

#91 Node GH rocks! @protoEvangelion (in 4 years)
"
`;

exports[`E2E: Issues Module Test Open Issue \`gh is --open --number 7 --number 10\` 1`] = `
"Opening issue 7 on protoEvangelion/gh
https://github.com/protoEvangelion/gh/issues/7
Opening issue 10 on protoEvangelion/gh
https://github.com/protoEvangelion/gh/issues/10
"
`;

exports[`E2E: Issues Module Test Open issue in browser \`gh is 55 --browser\` 1`] = `
"https://github.com/protoEvangelion/gh/issues/55
"
`;

exports[`E2E: Issues Module Test Search Issues \`gh is -s hi\` 1`] = `
"Searching for hi in issues for protoEvangelion/gh

#72 hi @protoEvangelion (in 4 years)

#65 hi @protoEvangelion (in 4 years)

#63 hi @protoEvangelion (in 4 years)

#62 hi @protoEvangelion (in 4 years)

#61 hi @protoEvangelion (in 4 years)

#60 hi @protoEvangelion (in 4 years)

#19 hi @protoEvangelion (in 4 years)

#1 test1 @protoEvangelion (in 4 years)
"
`;
